### PR TITLE
fix: Error when updating ParseObject with relation to another object

### PR DESCRIPTION
**Pull request by juriSacchetta**
### New Pull Request Checklist

- [X] I am not disclosing a [vulnerability](https://github.com/parse-community/Parse-SDK-Flutter/security/policy).
- [x] I am creating this PR in reference to an issue.

### Issue Description

Closes: https://github.com/parse-community/Parse-SDK-Flutter/issues/696

If you try to add a relationship to an object and after you do another operation to it you'll face this error: 

>Unhandled Exception: type '_InternalLinkedHashMap<String, dynamic>' is not a subtype of type 'ParseRelation<ParseObject>?' in type cast

By adding a relation, the set method of ParseObject changes the value corresponding to the key. Thus, the next time the relation is attempted, there will be no instance of ParseRelation but a Map.

```
final PostModel = ...;
final LikeModel like = LikeModel();

like.onPost = post;
await like.save();

post.likeRelation.add(like); // or post.addRelation('likes', [post]);
ParseResponse resp = await post.update();
if (resp.success) {
  post.likeRelation.remove(like); // Bad
}
```

### Approach

To solve this bug, I had to insert an if within the set method, but it is not a clean solution. I could not find a better solution. Honestly, I think using the same map to store both data and operations is a risky decision.

The last change concerns the creation of a relation in the client.
Every time the relation was accessed, a new one was created, thus also losing the cache function that was already implemented.
I tried to always return the same instance.

### TODOs before merging

- [x] Add tests
- [x] Add changes to documentation (guides, repository pages, in-code descriptions)
- [ ] A changelog entry

### CODE REVIEW

**Comment by mtrezza** [comment]
Could you please create an issue, move the explanation to that issue and just reference the issue in this PR?

**Comment by juriSacchetta** [comment]
It should work now 😁

Let me know!!

**Review by mbfakourii — DISMISSED** [review]
The code has been tested, it seems good.
@juriSacchetta is it possible to add the test?

**Comment by juriSacchetta** [comment]
At the moment, I didn't write a test but I've tested it inside my project. It seems to work fine.

Maybe in the next few days, I'll write one. I can't guarantee it, my schedule is very busy sorry.

**Comment by mtrezza** [comment]
We'll keep this open until a test is added.

**Comment by juriSacchetta** [comment]
Hey guys I added a new test relation specific. Let me know if it can be enough :)

**Comment by mtrezza** [comment]
Rebased and started CI

File: packages/dart/test/relation_test.dart [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+      like2.objectId = "like2";
+
+      like.onPost = post;
+      like2.onPost = post;
+
+      post.likeRelation.add(like); // or post.addRelation('likes', [post]);
+      post.likeRelation.add(like2); // or post.addRelation('likes', [post]);
+
+      // desired output
+      String expectedResult =
```
Comment (mbfakourii) [inline packages/dart/test/relation_test.dart:42]: Tests are not like other tests, please write like other samples

**Review by mbfakourii — CHANGES_REQUESTED** [review]
Tests are not like other tests, please write like other samples

File: packages/dart/test/relation_test.dart [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+          "{\"Likes\":{\"__op\":\"AddRelation\",\"objects\":[{\"__type\":\"Pointer\",\"className\":\"Like\",\"objectId\":\"like1\"},{\"__type\":\"Pointer\",\"className\":\"Like\",\"objectId\":\"like2\"}]}}";
```
Comment (mbfakourii) [inline packages/dart/test/relation_test.dart:43]: jsons are not written in a standard form

File: packages/dart/test/relation_test.dart [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    String expectedResult =
+        "{\"Likes\":{\"__op\":\"RemoveRelation\",\"objects\":[{\"__type\":\"Pointer\",\"className\":\"Like\",\"objectId\":\"like\"}]}}";
+    // act
+    dynamic actualResult = post.encode();
+    //assert
+    expect(actualResult.toString(), expectedResult);
+  });
+}
+
+class PostModel extends ParseObject implements ParseCloneable {
```
Comment (mbfakourii) [inline packages/dart/test/relation_test.dart:86]: What is the reason for creating this class?

File: packages/dart/test/relation_test.dart (hunk 2 above)
Comment (juriSacchetta) [reply packages/dart/test/relation_test.dart]: I fixed it

File: packages/dart/test/relation_test.dart (hunk 3 above)
Comment (juriSacchetta) [reply packages/dart/test/relation_test.dart]: I re-wrote the test without this class

File: packages/dart/test/relation_test.dart [hunk 4]
```diff
@@ -0,0 +1,101 @@
+import 'dart:convert';
+
+import 'package:parse_server_sdk/parse_server_sdk.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('Relation', () {
+    test('addRelation', () async {
+      // arrange
+      await Parse().initialize(
```
Comment (mbfakourii) [inline packages/dart/test/relation_test.dart:10]: This section does not need to be written twice, write it once in the setup section, it is enough to take a look at [this section](https://github.com/parse-community/Parse-SDK-Flutter/blob/b07ce8c7fbe709333a6a97ce5473e6d5c88f33f0/packages/dart/test/parse_query_test.dart#L12)

File: packages/dart/test/relation_test.dart [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        }
+      };
+
+      var act = parent.toJson(forApiRQ: true);
+      expect(act, expectedResult);
+    });
+  });
+  test('removeRelation', () async {
+    // arrange
+    await Parse().initialize(
```
Comment (mbfakourii) [inline packages/dart/test/relation_test.dart:64]: And this part too

**Review by mbfakourii — CHANGES_REQUESTED** [review]
The rest of the parts are fine, just move the duplicate codes to the setup section

File: packages/dart/test/relation_test.dart (hunk 4 above)
Comment (juriSacchetta) [reply packages/dart/test/relation_test.dart:10]: I'll do it right away, thank you very much for guiding me through the corrections, and thanks for your work

**Comment by xanderelsmith** [comment]
Hi , please i keep getting this issue "Unhandled Exception: type '_InternalLinkedHashMap<String, dynamic>' is not a subtype of type 'ParseRelation?' in type cast" 


how dp i fix it?

**Comment by mtrezza** [comment]
@xanderelsmith is this related to this PR?

**Comment by mbfakourii** [comment]
> Hi , please i keep getting this issue "Unhandled Exception: type '_InternalLinkedHashMap<String, dynamic>' is not a subtype of type 'ParseRelation?' in type cast"
> 
> how dp i fix it?

This issue is not related to this section. Please open an issue if you have a problem

**Comment by juriSacchetta** [comment]
Hey there!
I noticed that my branch has been merged, so can I close the PR? Thanks again :)

**Review by mbfakourii — APPROVED** [review]
It was tested and there is no problem

**Comment by mbfakourii** [comment]
> Hey there! I noticed that my branch has been merged, so can I close the PR? Thanks again :)

No, it has not been merged yet

**Comment by xanderelsmith** [comment]
> @xanderelsmith is this related to this PR?

Yes

**Comment by mtrezza** [comment]
@xanderelsmith How is it related to this PR? Could you give more details?

**Comment by xanderelsmith** [comment]
It throws an error, when removing a parse object from a parse relation

**Comment by xanderelsmith** [comment]
You said something about the data becoming a map once
newobject.remove() is done

**Comment by xanderelsmith** [comment]
> ### New Pull Request Checklist
> * [x]  I am not disclosing a [vulnerability](https://github.com/parse-community/Parse-SDK-Flutter/security/policy).
> * [x]  I am creating this PR in reference to an issue.
> 
> ### Issue Description
> Closes: #696
> 
> If you try to add a relationship to an object and after you do another operation to it you'll face this error:
> 
> > Unhandled Exception: type '_InternalLinkedHashMap<String, dynamic>' is not a subtype of type 'ParseRelation?' in type cast
> 
> By adding a relation, the set method of ParseObject changes the value corresponding to the key. Thus, the next time the relation is attempted, there will be no instance of ParseRelation but a Map.
> 
> ```
> final PostModel = ...;
> final LikeModel like = LikeModel();
> 
> like.onPost = post;
> await like.save();
> 
> post.likeRelation.add(like); // or post.addRelation('likes', [post]);
> ParseResponse resp = await post.update();
> if (resp.success) {
>   post.likeRelation.remove(like); // Bad
> }
> ```
> 
> ### Approach
> To solve this bug, I had to insert an if within the set method, but it is not a clean solution. I could not find a better solution. Honestly, I think using the same map to store both data and operations is a risky decision.
> 
> The last change concerns the creation of a relation in the client. Every time the relation was accessed, a new one was created, thus also losing the cache function that was already implemented. I tried to always return the same instance.
> 
> ### TODOs before merging
> * [x]  Add tests
> * [x]  Add changes to documentation (guides, repository pages, in-code descriptions)
> * [ ]  A changelog entry

This🤲

**Review by Nidal-Bakir — COMMENTED** [review]
Thanks for your contribution @juriSacchetta, I'm working to fix this issue and other issues related to using `Map`s to store the operations, I'm using a more OOP approach to improve and fix all of them using `ParseOperations`class, the `ParseMergeTool` class will be removed. I will definitely take a look at your code and see if I use any of it. Thanks again for your work.

Here is my PR: #860

@parse-community/flutter-sdk-review, I'd like to request that we don't merge this pull request just yet. Can we hold off for a bit? I'm working on a more comprehensive solution.

Thanks

**Comment by mtrezza** [comment]
Fine with me, @mbfakourii what do you think? 

Do you have an ETA for the other PR you're working on? If it takes longer, another possibility may be to merge this for a quick fix and mere the comprehensive fix later on.

**Comment by Nidal-Bakir** [comment]
I'm currently working on the `ParseRelation` and it is the last thing to do on my list of TODOs, then I will add tests and documntation for the functions and classes and then I will do any refactoring/cleaning when I look at it again.

I think I need a  week or less.

> another possibility may be to merge this for a quick fix and merge the comprehensive fix later on.

Sure, why not. Then give some time to review and test this PR

**Comment by Nidal-Bakir** [comment]
That is a breaking change!

**Comment by Nidal-Bakir** [comment]
Can you update your master branch to be in sync with upstream, so you can see the added tests? I just checked out your master and pulled the test files from upstream, and the tests are reporting syntax errors and failing.

**Comment by mbfakourii** [comment]
> Fine with me, @mbfakourii what do you think?
> 
> Do you have an ETA for the other PR you're working on? If it takes longer, another possibility may be to merge this for a quick fix and mere the comprehensive fix later on.

I think it is good and there is no problem

**Comment by juriSacchetta** [comment]
> Fine with me, @mbfakourii what do you think? 
> 
> Do you have an ETA for the other PR you're working on? If it takes longer, another possibility may be to merge this for a quick fix and mere the comprehensive fix later on.

Hi guys, I really need this fix as soon as possible. This solution would be perfect for me.
Please resolve fast this problem in any way you prefer. 

Thank you for your work 😁

**Comment by Nidal-Bakir** [comment]
>Hi guys, I really need this fix as soon as possible. This solution would be perfect for me.
Please resolve fast this problem in any way you prefer.

We are working on it!

You can just use the SDK as an embedded package in your project and import it using its path:./ in the `pubspec.yaml`
when you have the SDK embedded in your project you can just do and fix it as you wish.

see (https://docs.flutter.dev/development/packages-and-plugins/using-packages#dependencies-on-unpublished-packages)

**Comment by Nidal-Bakir** [comment]
Completed via #860

**Comment by xanderelsmith** [comment]
> > Hi guys, I really need this fix as soon as possible. This solution would be perfect for me.
> > Please resolve fast this problem in any way you prefer.
> 
> We are working on it!
> 
> You can just use the SDK as an embedded package in your project and import it using its `path: ./` in the `pubspec.yaml` when you have the SDK embedded in your project you can just do and fix it as you wish.
> 
> here is a code snippte:
> 
> ```yaml
>   parse_server_sdk_flutter:
>     path: ./Parse-SDK-Flutter/packages/flutter
> ```
> 
> see (https://docs.flutter.dev/development/packages-and-plugins/using-packages#dependencies-on-unpublished-packages)



Hi, please Can I use this,
Relation.remove doesn't work for me in the latest version

**Pull request closed without merging**